### PR TITLE
Feat: cross check scheduling

### PR DESCRIPTION
**Pull request by AlreadyBored**
**🟢 Add `deploy` label if you want to deploy this Pull Request to staging environment**

#### 🧑‍⚖️ Pull Request Naming Convention

- Title should follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- Do not put issue id in title
- Do not put WIP in title. Use [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) functionality
- Consider to add `area:*` label(s)

* [x] I followed naming convention rules

---

#### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Performance optimization
- [ ] Refactoring
- [x] Test Case
- [x] Documentation update
- [ ] Other

#### 🔗 Related issue link

https://github.com/rolling-scopes/rsschool-app/issues/1212
https://github.com/rolling-scopes/rsschool-app/issues/1044
https://github.com/rolling-scopes/rsschool-app/issues/431

#### 💡 Background and solution

Implemented functionality for starting Cross-Check automatically and scheduling its finish.
**NB! Before merging this PR we need to set correct status for existing course tasks!** 
It can be done by running SQL-script like
```sql
-- Update for old tasks completed before
UPDATE "course_task" AS c
SET "crossCheckStatus" = 'completed'
FROM "task_result" AS t
WHERE "c"."checker" = 'crossCheck'
AND "t"."score" != 0
AND (
"c"."studentEndDate" < CURRENT_DATE
OR
-- Nullish studentEndDate can be set on some old tasks
"c"."studentEndDate" = NULL
);

-- Update for tasks in progress
UPDATE "course_task" AS c
SET "crossCheckStatus" = 'distributed'
FROM "task_result" t
WHERE "c"."checker" = 'crossCheck'
AND "t"."score" = 0
AND "t"."score" IS DISTINCT FROM "t"."score"
AND "c"."studentEndDate" < CURRENT_DATE
AND "c"."studentEndDate" != NULL
```

#### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Database migration is added or not needed
- [x] Documentation is updated/provided or not needed
- [x] Changes are tested locally

### CODE REVIEW

File: server/src/models/courseTask.ts [hunk 1]
```diff
@@ -102,4 +107,7 @@ export class CourseTask {
 
   @Column({ nullable: true, type: 'int' })
   duration: number;
+
+  @Column({ type: 'varchar', default: 'initial' })
+  crossCheckState: CrossCheckState;
```
Comment (aaliakseyenka) [inline server/src/models/courseTask.ts:112]: consider using `status` instead of `state`

File: nestjs/src/cross-check/utils/constants.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export const DEV_HOST = 'http://localhost:3000';
+
+export const PROD_HOST = 'https://app.rs.school';
```
Comment (aaliakseyenka) [inline nestjs/src/cross-check/utils/constants.ts:3]: I guess this constants are not related to cross-check. Let's put them to `config.service`. Also by default we should be using localhost, if respective env variable is not set

File: nestjs/src/cross-check/utils/tasksFiltering.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { CourseTask } from '@entities/courseTask';
+
+export const getTaskFilters = () => {
+  const currTimestampUTC = Date.now();
```
Comment (aaliakseyenka) [inline nestjs/src/cross-check/utils/tasksFiltering.ts:4]: I am not sure this is good pattern:  since it is in closure there is no guarantee it will be correct value when called by fns below

File: nestjs/src/cross-check/cross-check.module.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { Module } from '@nestjs/common';
+import { TypeOrmModule } from '@nestjs/typeorm';
+import { HttpModule } from '@nestjs/axios';
+import { ScheduleModule } from '@nestjs/schedule';
+import { CourseTask } from '@entities/courseTask';
+import { ConfigModule } from '../config/config.module';
+import { CrossCheckService } from './cross-check.service';
+
+@Module({
+  imports: [ScheduleModule.forRoot(), TypeOrmModule.forFeature([CourseTask]), HttpModule, ConfigModule],
```
Comment (aaliakseyenka) [inline nestjs/src/cross-check/cross-check.module.ts:10]: shouldn't schedule be initialized on the app level?

File: nestjs/src/cross-check/cross-check.service.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    const { tasksToStart, tasksToFinish } = await this.getCrossCheckTasks();
+    await this.initCrossCheckAction(tasksToStart, 'distribution');
+    await this.initCrossCheckAction(tasksToFinish, 'completion');
+  }
+
+  private getInitalDataForRequest() {
+    const { username, password } = this.conf.users.root;
+
+    const authHeader = `Basic ${Buffer.from(username + ':' + password).toString('base64')}`;
+    const host = this.conf.isDev ? DEV_HOST : PROD_HOST;
```
Comment (aaliakseyenka) [inline nestjs/src/cross-check/cross-check.service.ts:34]: make sure to put this logic into config service itself

File: nestjs/src/cross-check/cross-check.service.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  private readonly logger = new Logger('CrossCheckService');
+
+  constructor(
+    @InjectRepository(CourseTask)
+    private readonly courseTaskRepository: Repository<CourseTask>,
+    private readonly httpService: HttpService,
+    private readonly conf: ConfigService,
+  ) {}
+
+  @Cron('5 0 * * *', { timeZone: 'UTC' })
```
Comment (aaliakseyenka) [inline nestjs/src/cross-check/cross-check.service.ts:23]: wondering is needed to call this job so often?

Comment (AlreadyBored) [reply nestjs/src/cross-check/cross-check.service.ts:23]: This means means "every day once a day at 00:05 UTC". This is the lowest possible frequency for this operation, as we have to update the CourseTask status daily.
image

Comment (aaliakseyenka) [reply nestjs/src/cross-check/cross-check.service.ts:23]: I am didn't check, but according to documentation https://docs.nestjs.com/techniques/task-scheduling first value is second

File: nestjs/src/cross-check/utils/tasksFiltering.ts (hunk 3 above)
Comment (AlreadyBored) [reply nestjs/src/cross-check/utils/tasksFiltering.ts:4]: To be honest, this solution seems pretty good to me. Storing the value in a closure allows you to avoid unnecessary time calculations, and also isolates the received time value from external access. But perhaps I missed something, could you please give an example of a problem that may arise here?

File: nestjs/src/cross-check/cross-check.service.ts (hunk 6 above)
Comment (AlreadyBored) [reply nestjs/src/cross-check/cross-check.service.ts:23]: It's optional :)
If you don't provide seconds, it starts with minutes

File: nestjs/src/cross-check/utils/tasksFiltering.ts (hunk 3 above)
Comment (aaliakseyenka) [reply nestjs/src/cross-check/utils/tasksFiltering.ts:4]: while i agree about complex calculations, taking current time is not expensive( literally nothing) and u save current time. at the moment when the function is called the time might be changed significantly. there is no problem in you current code. but this is a good place for confusion and future errors

File: nestjs/src/cross-check/cross-check.service.ts (hunk 6 above)
Comment (aaliakseyenka) [reply nestjs/src/cross-check/cross-check.service.ts:23]: understood, didn't pay attention to number of symbols)

File: client/src/components/Table/renderers.tsx [hunk 7]
```diff
@@ -9,13 +9,29 @@ import {
...
+import { Checker, CrossCheckStatus } from '../../services/course';
 
 const { Text } = Typography;
 
 export function dateRenderer(value: string | null) {
   return value ? moment(value).format('YYYY-MM-DD') : '';
 }
 
+export function crossCheckDateRenderer(value: string | null, { checker }: { checker: Checker }) {
+  if (checker !== 'crossCheck') return 'NA';
```
Comment (mikhama) [inline client/src/components/Table/renderers.tsx:21]: Do you mean `N/A`?

File: client/src/components/Table/renderers.tsx [hunk 8]
```diff
@@ -9,13 +9,29 @@ import {
...
+  return value ? moment(value).tz('UTC').format('YYYY-MM-DD') : 'Not Set';
+}
+
+export function crossCheckStatusRenderer(value: CrossCheckStatus, { checker }: { checker: Checker }) {
+  return checker !== 'crossCheck' ? (
+    'NA'
```
Comment (mikhama) [inline client/src/components/Table/renderers.tsx:27]: `N/A`

File: client/src/services/course.ts [hunk 9]
```diff
@@ -10,7 +10,8 @@ import { PreferredStudentsLocation } from 'common/enums/mentor';
 import { CrossCheckFieldsTypes } from '../pages/course/admin/cross-check-table';
 import { CoursesTasksApi } from 'api';
 
-type Checker = 'auto-test' | 'mentor' | 'assigned' | 'taskOwner' | 'crossCheck';
+export type CrossCheckStatus = 'initial' | 'distributed' | 'completed';
+export type Checker = 'auto-test' | 'mentor' | 'assigned' | 'taskOwner' | 'crossCheck';
```
Comment (mikhama) [inline client/src/services/course.ts:14]: Can you chose only one case: kebab or camel?

File: client/src/services/course.ts [hunk 10]
```diff
@@ -25,6 +26,9 @@ export interface CourseTask {
   descriptionUrl: string | null;
   studentStartDate: string | null;
   studentEndDate: string | null;
+  crossCheckEndDate: string | null;
+  crossCheckStatus: CrossCheckStatus;
+  useJury: boolean;
```
Comment (mikhama) [inline client/src/services/course.ts:31]: I think it's not better naming for react ecosystem. I will expect hook with such name.
Can you rename it in something like `isJuryInvolved` or `isCheckedByJury`?

File: nestjs/src/config/config.service.ts [hunk 11]
```diff
@@ -35,12 +35,16 @@ type AWSServices = {
   accessKeyId: string;
 };
 
+const DEV_HOST = 'http://localhost:3000';
+const PROD_HOST = 'https://app.rs.school';
```
Comment (mikhama) [inline nestjs/src/config/config.service.ts:39]: Isn't it better to take them from the env file?

File: nestjs/src/cross-check/cross-check.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  describe('executeCronJobs should work correctly', () => {
+    beforeAll(async () => {
+      jest.spyOn(Date, 'now').mockImplementation(() => MOCK_CURRENT_TIMESTAMP);
+      await service.executeCronJobs();
+    });
+
+    it('httpService.post should be called right amout of times', () => {
+      const { completion, distribution } = expectedResults.expectedHttpServiceCallsAmount;
+      const { length: totalCallsAmount } = mockPost.mock.calls;
+      expect(totalCallsAmount).toBe(completion + distribution);
```
Comment (mikhama) [inline nestjs/src/cross-check/cross-check.service.spec.ts:59]: expect(mockPost).toBeCalledTimes(completion + distribution);

File: nestjs/src/cross-check/cross-check.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    });
+
+    it('tasks with submission deadline passed should be distributed', () => {
+      const distributionUrls = mockPost.mock.calls
+        .map(([url]) => url)
+        .filter((url: string) => url.endsWith('distribution'));
+      expect(distributionUrls).toEqual(expectedResults.expectedRequestsUrl.distribution);
```
Comment (mikhama) [inline nestjs/src/cross-check/cross-check.service.spec.ts:66]: mockPost.toBeCalledWith(..?

File: nestjs/src/cross-check/cross-check.service.spec.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    });
+
+    it('tasks with cross-check deadline passed should be completed', () => {
+      const distributionUrls = mockPost.mock.calls
+        .map(([url]) => url)
+        .filter((url: string) => url.endsWith('completion'));
+      expect(distributionUrls).toEqual(expectedResults.expectedRequestsUrl.completion);
```
Comment (mikhama) [inline nestjs/src/cross-check/cross-check.service.spec.ts:73]: toBeCalledWith

File: nestjs/src/cross-check/utils/test-utils.ts [hunk 15]
```diff
@@ -0,0 +1,128 @@
+// Date is equal to 2022-03-22 00:05 UTC
+export const MOCK_CURRENT_TIMESTAMP = 1647907500000;
```
Comment (mikhama) [inline nestjs/src/cross-check/utils/test-utils.ts:2]: export const MOCK_CURRENT_TIMESTAMP = `new Date('2022-03-22 00:05 UTC').getTime();` => [S] - Self-descriptable code 😄

File: nestjs/src/cross-check/utils/tasksFiltering.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import { CourseTask } from '@entities/courseTask';
```
Comment (mikhama) [inline nestjs/src/cross-check/utils/tasksFiltering.ts:1]: `tasksFiltering.ts` - I thought we use kebab-case everywhere. Am I not right?

File: nestjs/src/config/config.service.ts (hunk 11 above)
Comment (aaliakseyenka) [reply nestjs/src/config/config.service.ts:39]: i think current way is not what we meant: the host itself shouldn't be in the code. it should be in .env, ie the url itself should be set in the env variable, not boolean value

File: nestjs/package.json [hunk 17]
```diff
@@ -32,6 +32,7 @@
     "@nestjs/core": "8.4.3",
     "@nestjs/mapped-types": "1.0.1",
     "@nestjs/passport": "8.2.1",
+    "@nestjs/schedule": "^1.0.2",
```
Comment (apalchys) [inline nestjs/package.json:35]: please remove `^`

File: nestjs/src/config/config.service.ts [hunk 18]
```diff
@@ -76,5 +81,12 @@ export class ConfigService {
       hirers: conf.get('RSSHCOOL_USERS_HIRERS')?.split(',') ?? [],
       admins: conf.get('RSSHCOOL_USERS_ADMINS')?.split(',') ?? [],
     };
+
+    const DEV_HOST = conf.get('RSSCHOOL_CROSS_CHECK_SCHEDULING_DEV_HOST') ?? '';
```
Comment (apalchys) [inline nestjs/src/config/config.service.ts:85]: this looks confusing. Why do we need separate config values for `prod` and `dev`? and what's about e.g. feature branches ?

Comment (apalchys) [reply nestjs/src/config/config.service.ts:85]: if you just need host, please see `RSSHCOOL_HOST`  and you can port it to `nestjs`. I would avoid creating a new config values

File: nestjs/src/cross-check/cross-check.service.spec.ts [hunk 19]
```diff
@@ -0,0 +1,189 @@
+import { HttpService } from '@nestjs/axios';
```
Comment (apalchys) [inline nestjs/src/cross-check/cross-check.service.spec.ts:1]: 👍

File: nestjs/src/cross-check/cross-check.service.ts [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
```
Comment (apalchys) [inline nestjs/src/cross-check/cross-check.service.ts:22]: Cron is tricky to read for newcomers. I would add comment or self descriptive variable to explain what `5 0 * * *` means

File: nestjs/src/cross-check/cross-check.service.ts [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+  async executeCronJobs() {
+    const { tasksToStart, tasksToFinish } = await this.getCrossCheckTasks();
+    await this.initCrossCheckAction(tasksToStart, 'distribution');
+    await this.initCrossCheckAction(tasksToFinish, 'completion');
+  }
+
+  private getInitalDataForRequest() {
+    const { username, password } = this.conf.users.root;
+
+    const authHeader = `Basic ${Buffer.from(username + ':' + password).toString('base64')}`;
```
Comment (apalchys) [inline nestjs/src/cross-check/cross-check.service.ts:32]: Please use built-in axios functionality instead of generating  manually. See `https://github.com/axios/axios`

```
// `auth` indicates that HTTP Basic auth should be used, and supplies credentials.
  // This will set an `Authorization` header, overwriting any existing
  // `Authorization` custom headers you have set using `headers`.
  // Please note that only HTTP Basic auth is configurable through this parameter.
  // For Bearer tokens and such, use `Authorization` custom headers instead.
  auth: {
    username: 'janedoe',
    password: 's00pers3cret'
  },
```

File: nestjs/src/cross-check/utils/tasks-filtering.ts [hunk 22]
```diff
@@ -0,0 +1,13 @@
+import { CourseTask } from '@entities/courseTask';
```
Comment (apalchys) [inline nestjs/src/cross-check/utils/tasks-filtering.ts:1]: Please, no `utils` folders. Just put this file into `cross-check` folder.

File: nestjs/src/config/config.service.ts (hunk 18 above)
Comment (AlreadyBored) [reply nestjs/src/config/config.service.ts:85]: replaced wth `RSSCHOOL_HOST`

File: nestjs/src/cross-check/cross-check.service.spec.ts (hunk 19 above)
Comment (AlreadyBored) [reply nestjs/src/cross-check/cross-check.service.spec.ts:1]: What do you mean? It's used as value to be replaced by mock

**Review by mikhama — APPROVED** [review]

**Review by apalchys — APPROVED** [review]
